src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php

```diff
@@ -724,7 +724,9 @@ private function addNewInstance($id, Definition $definition, $return, $instantia
             }
 
             if (null !== $definition->getFactoryService()) {
-                return sprintf("        $return{$instantiation}%s->%s(%s);\n", $this->getServiceCall($definition->getFactoryService()), $definition->getFactoryMethod(), implode(', ', $arguments));
+                $service = $this->dumpValue($definition->getFactoryService());
+
+                return sprintf("        $return{$instantiation}%s->%s(%s);\n", 0 === strpos($service, '$') ? sprintf('$this->get(%s)', $service) : $this->getServiceCall($definition->getFactoryService()), $definition->getFactoryMethod(), implode(', ', $arguments));
```

Comment:
do we really need this logic here ? I don't understand the goal of the ``$`` check changing the way to dump things. ``$definition->getFactoryService()`` will always return a service id here, given you resolve it in the compiler pass